src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -113,6 +143,31 @@ public static Type InterfaceType
         /// </summary>
         public static event RefreshEventHandler? Refreshed;
 
+        private static readonly bool s_requireRegisteredTypes =
+            AppContext.TryGetSwitch(
+                switchName: "System.ComponentModel.TypeDescriptor.RequireRegisteredTypes",
+                isEnabled: out bool isEnabled)
+            ? isEnabled : false;
```

Comment:
Is this switch only used for the the existing APIs (without the `FromRegisteredType` suffix)? If yes, I assume the new APIs (with the `FromRegisteredType` suffix) will always check that the type is registered first.

Reply:
The "FromRegisteredType" APIs enforce the registration without the switch, at least for the normal cases like `GetPropertiesFromRegisteredType()` where the check more or less comes for free. Some "FromRegisteredType" APIs, to avoid an unnecessary perf hit, don't (see [this](https://github.com/dotnet/runtime/pull/102094/files#diff-8cc7ee2f29d23a07e206917dfbaab6311a6dbc40c7ea6cb09f120780a3375181R104-R105)).

When the switch is on:
- The existing APIs (e.g. `TypeDescriptor.GetProperties()`) check to see if the type is registered for the internal reflection provider. This allows existing code to basically remain as-is (for compat with code that may not be owned) but does require registering the type. This avoids issues where there are likely suppressed trim warning previously. Previously we discussed always throwing for these existing methods, but I thought that was too extreme.
- The base class implementation of "FromRegisteredType()" APIs check the `bool? RequiresRegistration` property which defaults to return `null`. These will now throw `NotImplementedException` when the switch is on. For example, see [this](https://github.com/dotnet/runtime/pull/102094/files#diff-d3767bcbab3563f1f0d10d6d3e56c42bd6f54d1565a4b8195c6feba8012a7997R234). This enforces that providers need to be updated to declare whether or not they require registration if the "FromRegisteredType" APIs are used - most or all providers will return `false`; `true` is only needed to be returned if reflection is used at some point on that type. Existing APIs will **not** throw `NotImplementedException`.